### PR TITLE
1182 handle direct lambda invocation errors

### PR DESCRIPTION
**Pull request by leite08**
Ref: https://github.com/metriport/metriport-internal/issues/1182

### Dependencies

- upstream: https://github.com/metriport/metriport/pull/1022
- downstream: none

### Description

- Fix how we handle direct lambda invocation errors.
- Make lambdas have shared signatures (input/output types) shared w/ their clients.

### Release Plan

- staging
   - [x] release NPM packages
   - [x] update lambda
   - [ ] merge this
   - [ ] test
      - [ ] document query
      - [ ] download XML
      - [ ] download PDF
      - [ ] download HTML
      - [ ] visualize PDF
      - [ ] visualize HTML
      - [ ] `GET /consolidated` w/ conversion (checking result on WH call):
         - [ ] `xml`
         - [ ] `pdf`
         - [ ] `html`
- production
   - [ ] release NPM packages
   - [ ] update lambda
   - [ ] merge this

### CODE REVIEW

File: packages/core/src/external/aws/lambda.ts [hunk 1]
```diff
@@ -9,3 +11,88 @@ export function logResultToString(logResult: string | undefined): string | undef
...
+}: {
+  result: PromiseResult<AWS.Lambda.InvocationResponse, AWS.AWSError>;
+  lambdaName?: string;
+  failGracefuly?: boolean;
+  log?: typeof console.log;
+}): string | undefined {
+  if (result.StatusCode !== 200) {
+    if (!failGracefuly) {
+      throw new MetriportError("Lambda invocation failed", undefined, { lambdaName });
+    } else return undefined;
```
Comment (RamilGaripov) [inline packages/core/src/external/aws/lambda.ts:83]: nit: dont need the else

File: packages/core/src/external/aws/lambda.ts [hunk 2]
```diff
@@ -9,3 +11,88 @@ export function logResultToString(logResult: string | undefined): string | undef
...
+  }
+  if (!result.Payload) {
+    if (!failGracefuly) {
+      throw new MetriportError("Lambda payload is undefined", undefined, { lambdaName });
+    } else return undefined;
```
Comment (RamilGaripov) [inline packages/core/src/external/aws/lambda.ts:88]: nit: ditto

File: packages/core/src/external/aws/lambda.ts [hunk 3]
```diff
@@ -9,3 +11,88 @@ export function logResultToString(logResult: string | undefined): string | undef
...
+  }
+  if (isLambdaError(result)) {
+    const msg = `Error calling lambda ${lambdaName}`;
+    const errorDetails = JSON.stringify(getLambdaError(result));
+    log(`${msg} - ${errorDetails}`);
+    if (!failGracefuly) throw new MetriportError(msg, undefined, { lambdaName, errorDetails });
+    else return undefined;
```
Comment (RamilGaripov) [inline packages/core/src/external/aws/lambda.ts:95]: nit: ditto

**Review by RamilGaripov — APPROVED** [review]
Approved and left a couple of tiny nits

File: packages/core/src/external/aws/lambda.ts (hunk 1 above)
Comment (leite08) [reply packages/core/src/external/aws/lambda.ts:83]: 💯

File: packages/core/src/external/aws/lambda.ts [hunk 4]
```diff
@@ -9,3 +11,86 @@ export function logResultToString(logResult: string | undefined): string | undef
...
+  lambdaName?: string;
+  failGracefuly?: boolean | false;
+  log?: typeof console.log;
+}): string;
+export function getLambdaResultPayload(params: {
+  result: PromiseResult<AWS.Lambda.InvocationResponse, AWS.AWSError>;
+  lambdaName?: string;
+  failGracefuly: true;
+  log?: typeof console.log;
+}): string | undefined;
```
Comment (Goncharo) [inline packages/core/src/external/aws/lambda.ts:68]: why are these 2 exports needed if there's the one below?

File: packages/api/src/external/commonwell/document/document-query.ts [hunk 5]
```diff
@@ -752,23 +753,26 @@ async function triggerDownloadDocument({
...
+  //
+  // This should show up on a conflict w/ this PR https://github.com/metriport/metriport/pull/1013
+  // Need to make sure the cloud implentation of the doc downloader uses the new logic to handle lambda errors
+  //
+  // This should show up on a conflict w/ this PR https://github.com/metriport/metriport/pull/1013
+  // Need to make sure the cloud implentation of the doc downloader uses the new logic to handle lambda errors
+  //
+  // This should show up on a conflict w/ this PR https://github.com/metriport/metriport/pull/1013
+  // Need to make sure the cloud implentation of the doc downloader uses the new logic to handle lambda errors
+  //
```
Comment (Goncharo) [inline packages/api/src/external/commonwell/document/document-query.ts:772]: does the release plan need to be updated to take this into account?

**Review by Goncharo — APPROVED** [review]
LGTM - couple of questions

File: packages/core/src/external/aws/lambda.ts (hunk 4 above)
Comment (leite08) [reply packages/core/src/external/aws/lambda.ts:68]: Alloes for calls w/ diff params to have diff response types

File: packages/api/src/external/commonwell/document/document-query.ts (hunk 5 above)
Comment (leite08) [reply packages/api/src/external/commonwell/document/document-query.ts:772]: I'll have to fix this either here or on the PR that makes introduces the local impl, depending on which comes first.

I can choose one if both are g2g and let you know about the merge fix prior to merging it

**Pull request merged**